apps/script/HoC2023AiGenerateWeights.py

```diff
@@ -107,41 +111,91 @@
 background_dict = {}
 foreground_dict = {}
 
-# Calculate and print similarity scores
-for emoji_name in emoji_names_list:
-    palette_scores = []
-    for palette_word in palette_model_descriptive_names:
-        id_token = nlp(emoji_name)
-        palette_token = nlp(palette_word)
-        similarity_score = id_token.similarity(palette_token)
-        palette_scores.append(round(similarity_score, 2))
+# Define caches to store ada's raw embedding outputs to reduce query costs
+# These caches are stored as pickle files; python's native way to serialize data
+embedding_paths = ['apps/static/dance/ai/model/input_embeddings.pkl',
+'apps/static/dance/ai/model/palette_embeddings.pkl',
+'apps/static/dance/ai/model/background_embeddings.pkl',
+'apps/static/dance/ai/model/foreground_embeddings.pkl']
+input_path = embedding_paths[0]
+palette_path = embedding_paths[1]
+background_path = embedding_paths[2]
+foreground_path = embedding_paths[3]
+
+def load_embedding_cache(path):
+    # load the cache if it exists, and save a copy to disk
+    try:
+        embedding_cache = pd.read_pickle(path)
+    except FileNotFoundError:
+        embedding_cache = {}
+    with open(path, "wb") as embedding_cache_file:
+        pickle.dump(embedding_cache, embedding_cache_file)
+    return embedding_cache
+
+caches = [load_embedding_cache(path) for path in embedding_paths]
+input_cache = caches[0]
+palette_cache = caches[1]
+background_cache = caches[2]
+foreground_cache = caches[3]
+
+# Define a function to retrieve embeddings from the cache if present, and otherwise request via the API
+def retrieve_embedding(string: str,
+    cache_path: str,
+    embedding_cache,
+    model: str = EMBEDDING_MODEL,
+) -> list:
+    """Return embedding of given string, using a cache to avoid recomputing."""
+    if (string, model) not in embedding_cache.keys():
+        embedding_cache[(string, model)] = get_embedding(string, model)
+        with open(cache_path, "wb") as embedding_cache_file:
+            pickle.dump(embedding_cache, embedding_cache_file)
+    return embedding_cache[(string, model)]
+
+# Retrieve embeddings for input emojis, palettes, backgrounds, and foregrounds in that order
+option_lists = [emoji_names_list, palette_model_descriptive_names, background_effect_model_descriptive_names, foreground_effect_model_descriptive_names]
+embeddings = []
+# Final output should be list of lists structured as [[[emoji1], [emoji2], ...], [[palette1]...]...]
+# Where embeddings[0] = all emoji embeddings, [1] = palettes, [2] = background, [3] = foreground
+for i in range (0, 4):
```

Comment:
It does feel like generally there are lots of places where we assume that values across many lists with the same index represent are associated. Probably out of scope to change that right now but generally feels like it'd be nice to have all of the associated data for a given type in a single structure (something like this in JS, might be trickier in Python):

emoji: {
  cache_path: "apps/static/dance/ai/model/input_embeddings.pkl",
  embedding_cache: load_embedding_cache("apps/static/dance/ai/model/input_embeddings.pkl")
  option_list: emoji_names_list,
  embeddings: ...
}

Reply:
This definitely makes sense! I agree it's out of scope right now, but it would make things more simple.